feat(Randomizer) checks for existing values

MikeActually opened the pull request:
added some checks to the randomizer to basically use a reduced array of values if every possible value hasn't yet been used
does not play friendly with async; whereas, the faster you type, th emore likely a duplicate is to occur before exhausting all available options

MikeActually commented on src/components/StackNameGenerator.jsx:
```diff
@@ -22,10 +22,11 @@ class StackNameGenerator extends React.Component {
-      return filteredResults[randomIndex];
+      const results = response.data.split("\n");
+      const filteredResults = results.filter(word => word != null && word !== "");
+      const deDupeResults = filteredResults.filter(word => !this.state.generatedList.includes(word));
```
reduces list of values to ones not yet used; if size is greater than 0, uses it for base for randomizer; if size is 0, simply uses the full list

DonutsInBelly replied:
This doesn't totally work because at this state in the runtime, the `this.state.generatedList` != `listOfThings` because `this.state.generatedList` refers to that last word that was produced, not the current one being built. You might be able to see that in this build: 
https://deploy-preview-5--condescending-heisenberg-6c385d.netlify.com/

MikeActually replied:
While that is true that the words are not the same, it doesn't truly cause any problems, and, may even result in a desired experience. Consider this, "Bike" results in one stack, changing the B to an M results in the same output, except that the first item is swapped.

I suppose it really comes down to what the desired result is.

I'll also take a look at the other suggestion at my earliest opportunity.

JamieDixon commented:
Would this change result in dupe words being completely omitted? I ask because I had a similar thought when I built https://codesandbox.io/s/j3x11xmjq5 which is heavily inspired by this app with some adjustments.

I opted to keep looking for unique values so if your name is "bob" you wouldn't end up with 2 words, you'd end up with 3 words but the `b`s would be unique.

The function looks like this:

```js
const pickUniqueOptions = (items = [], picked = [], depth = 0) => {
  const [head = [], ...tail] = items;
  const item = head[Math.floor(Math.random() * head.length)];
  const seenBefore = picked.includes(item);
  const maxTriesPerLetter = 10;

  if (
    // We've tried too many times, let's just return the value we found
    (seenBefore && depth >= maxTriesPerLetter) ||
    // Great! We found a unique value but there's more values to process yet. Let's carry on.
    (!seenBefore && tail.length)
  ) {
    return pickUniqueOptions(tail, [...picked, item], 0);
  }

  // Terminal case! Woo. Let's return our completed list.
  if (!seenBefore && !tail.length) {
    return [...picked, item];
  }

  // We've seen this word, let's try again and hope for a unique one!
  return pickUniqueOptions(items, picked, depth + 1);
};
```

DonutsInBelly commented:
@JamieDixon Thats a fairly clever solution, it took me a bit to try and understand what it was doing but I think it should work. I think the only thing I'm unsure about is why a depth and maxTries might be necessary; the amount of names isn't too large at the moment.

I think something like a combination of using that function to filter out duplicates and comparing the length on line 28 of this PR might be a good idea. The app should try its best to filter out words that have already been used, but if someone puts 25 b's into the input, it should still output words.

MikeActually commented:
@JamieDixon that's what the filter is there for, also, the filter is deterministic and doesn't leave to chance drawing the same value all 10 times. Lastly, the filter would allow for even 100 possible options for a letter, it would take 101 entries of that letter before seeing a duplicate

MikeActually committed:
feat(Randomizer) checks for existing values

	fixed react warning in App.jsx for target=blank
	added logic to retry list of strings so that dupes happen even less frequently
	added logic to retain values for the first X values, if they match the last build

MikeActually commented on src/App.jsx:
```diff
@@ -124,6 +124,7 @@ class App extends React.Component {
               <a
                 href="https://github.com/dominusbelli/whatthefuckismystack"
                 target="_blank"
+                rel="noopener noreferrer"
```
was causing a react warning and the change is harmless

MikeActually commented on src/components/StackNameGenerator.jsx:
```diff
@@ -16,29 +16,39 @@ class StackNameGenerator extends React.Component {
-      let filteredResults = results.filter(word => word != null && word !== "");
-      const randomIndex = Math.floor(Math.random() * filteredResults.length);
-      return filteredResults[randomIndex];
+      const results = this.dedupeLoop(1,response.data.split('\n').filter(Boolean),currentList);
```
runs dedupe subprocess and returns a value from the list created

MikeActually commented on src/components/StackNameGenerator.jsx:
```diff
@@ -16,29 +16,39 @@ class StackNameGenerator extends React.Component {
     });
   }
 
+  dedupeLoop(loopCount, possibleStrings, currentList) {
```
attempts to find values based on how many times the list has been iterated; fir example, first iteration will only grab values that have been used < 1 times; second iteration, less than 2, and so on; basically ensuring that when duplicates are encountered, a value will not be repeated until all available items have been exhausted for that iteration... furthermore, if you have 16 A's in the name and 4 possible A values, you are guaranteed to see each value exactly 4 times

MikeActually commented on src/components/StackNameGenerator.jsx:
```diff
@@ -16,29 +16,39 @@ class StackNameGenerator extends React.Component {
   async generateStack(stackName) {
-    let listOfThings = [];
-    for (var i = 0; i < stackName.length; i++) {
-      if(this.state.generatedList.length > i && this.state.generatedList[i].charAt(0).toLowerCase() === stackName.charAt(i).toLowerCase()) {
```
refactored to list normalization

MikeActually commented on src/components/StackNameGenerator.jsx:
```diff
@@ -16,29 +16,39 @@ class StackNameGenerator extends React.Component {
+    const prevStack = this.state.stackName;
+    let max = 0;
+    
+    while((max < (stackName.length <= prevStack.length ? stackName.length : prevStack.length)) && stackName[max] === prevStack[max]) {
```
determines how much of the start of the new string matches the old string and retains those values from last build

MikeActually commented on src/components/StackNameGenerator.jsx:
```diff
@@ -58,9 +68,9 @@ class StackNameGenerator extends React.Component {
+      stackNameList = this.state.generatedList.map((name, index) => {
         return (
-          <ListItem>
+          <ListItem key={`${name}-${index}`}>
```
react sanity

MikeActually commented:
@dominusbelli @JamieDixon -- updated PR to reflect implementation of discussed ideas; a bit more extensive, but really addresses limiting repeats
